markdown.lua

```diff
@@ -1,227 +1,134 @@
-    [11] = 'Opal',
-    [12] = 'Obsidian',
-}
-
+-- Utility functions
 local function getFileHandle()
-    return io.open(filename, writemode)
+    return assert(io.open(filename, writemode), "Error opening file: " .. filename)
 end
 
+
 local function closeFileHandle(handle)
-    handle:write('\n***\n\n')
+    handle:write('\n---\n\n')
     handle:close()
-    print ('Data exported to "' .. filename .. '"')
-end
-
-local function reformat(strin)
-    local strout = strin
-
-    -- [P] tags seem to indicate a new paragraph
-    local newline_idx = string.find(strout, '[P]', 1, true)
-    while newline_idx ~= nil do
-        strout = string.sub(strout, 1, newline_idx - 1) .. '\n***\n\n' .. string.sub(strout, newline_idx + 3)
-        newline_idx = string.find(strout, '[P]', 1, true)
-    end
-
-    -- [R] tags seem to indicate a new 'section'. Let's mark it with a horizontal line.
-    newline_idx = string.find(strout, '[R]', 1, true)
-    while newline_idx ~= nil do
-        strout = string.sub(strout, 1, newline_idx - 1) .. '\n***\n\n' .. string.sub(strout,newline_idx + 3)
-        newline_idx = string.find(strout, '[R]', 1, true)
+    if writemode == 'a' then
+        print('\nData appended to "' .. filename .. '"')
+    elseif writemode == 'w' then
+        print('\nData overwritten in "' .. filename .. '"')
     end
+end
 
-    -- No idea what [B] tags might indicate. Just removing them seems to work fine
-    newline_idx = string.find(strout, '[B]', 1, true)
-    while newline_idx ~= nil do
-        strout = string.sub(strout, 1, newline_idx - 1) .. string.sub(strout,newline_idx + 3)
-        newline_idx = string.find(strout, '[B]', 1, true)
-    end
+local function reformat(str)
+    -- [B] tags seem to indicate a new paragraph
+    -- [R] tags seem to indicate a sub-blocks of text.Treat them as paragraphs.
+    -- [P] tags seem to be redundant
+    -- [C] tags indicate color. Remove all color information
+    return str:gsub('%[B%]', '\n\n')
+              :gsub('%[R%]', '\n\n')
+              :gsub('%[P%]', '')
+              :gsub('%[C:%d+:%d+:%d+%]', '')
+              :gsub('\n\n+', '\n\n')
+end
 
-    -- Reddit doesn't support custom colors in markdown. We need to remove all color information :(
-    local color_idx = string.find(strout, '[C:', 1, true)
-    while color_idx ~= nil do
-        strout = string.sub(strout, 1, color_idx - 1) .. string.sub(strout, color_idx + 9)
-        color_idx = string.find(strout, '[C:', 1, true)
-    end
+local function getNameRaceAgeProf(unit)
+    --%s is a placeholder for a string, and %d is a placeholder for a number.
+    return string.format("%s, %d years old %s.", dfhack.units.getReadableName(unit), df.global.cur_year - unit.birth_year, dfhack.units.getProfessionName(unit))
+end
 
-    return strout
+-- Main logic for item and unit processing
+local item = dfhack.gui.getSelectedItem(true)
+local unit = dfhack.gui.getSelectedUnit(true)
+
+if not item and not unit then
+    print([[
```

Comment:
use `dfhack.printerr` for a nice red color

Reply:
This code returns white text. Is it expected behavior? 
![Screenshot 2023-11-23 204134](https://github.com/DFHack/scripts/assets/98843804/cfc00fb0-c7cb-4583-bf50-9733c1285f4f)

```
if not item and not unit then
    dfhack.printerr([[
Error: No unit or item is currently selected.
- To select a unit, click on it.
- For items that are installed as buildings (like statues or beds),
open the building's interface and click the magnifying glass icon.
Please select a valid target and try running the script again.]])
    -- Early return to avoid proceeding further if no unit or item is selected
    return
end
```

Reply:
yes, the red color only shows up on the DFHack console. the launcher output won't ever show color since color information isn't available via the API that `gui/launcher` uses

Reply:
Can I resolve this?

Reply:
I think the best way to resolve this would be to create a proper UI for the tool -- add an export link to the screens where export is supported. In fact, this might be a better solution than the global keybinding since the feature will be more discoverable.

Reply:
example overlay code:
```lua
--@ module=true

local overlay = require('plugins.overlay')
local widgets = require('gui.widgets')

MarkdownOverlay = defclass(MarkdownOverlay, overlay.OverlayWidget)
MarkdownOverlay.ATTRS{
    desc='Adds links to screen where text can be exported to a file.',
    default_pos={x=7, y=21},
    default_enabled=true,
    viewscreens={
        'dwarfmode/ViewSheets/ITEM',
        'dwarfmode/ViewSheets/UNIT',
    },
    frame={w=21, h=1},
}

function MarkdownOverlay:init()
    self:addviews{
        widgets.TextButton{
            view_id='button',
            label='Export text',
            key='CUSTOM_CTRL_T',
            on_activate=self:callback('do_export'),
        },
        widgets.TextButton{
            view_id='button_good',
            label='Export done',
            text_pen=COLOR_GREEN,
            key='CUSTOM_CTRL_T',
            visible=false,
        },
    }
end

function MarkdownOverlay:do_export()
    dfhack.run_script('markdown')
    self.subviews.button.visible = false
    self.subviews.button_good.visible = true
    local end_ms = dfhack.getTickCount() + 5000
    local function label_reset()
        if dfhack.getTickCount() < end_ms then
            dfhack.timeout(10, 'frames', label_reset)
        else
            self.subviews.button_good.visible = false
            self.subviews.button.visible = true
        end
    end
    label_reset()
end

OVERLAY_WIDGETS = {overlay=MarkdownOverlay}
```